// ng2-smart-table/esm5/lib/components/cell/cell-edit-mode/edit-cell-default.js
import { __decorate, __metadata } from "tslib";
import { Component, Output, EventEmitter, Input } from '@angular/core';
import { Cell } from '../../../lib/data-set/cell';
var EditCellDefault = /** @class */ (function () {
    function EditCellDefault() {
        this.inputClass = '';
        this.edited = new EventEmitter();
    }
    EditCellDefault.prototype.onEdited = function (event) {
        this.edited.next(event);
        return false;
    };
    EditCellDefault.prototype.onStopEditing = function () {
        this.cell.getRow().isInEditing = false;
        return false;
    };
    EditCellDefault.prototype.onClick = function (event) {
        event.stopPropagation();
    };
    __decorate([
        Input(),
        __metadata("design:type", Cell)
    ], EditCellDefault.prototype, "cell", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], EditCellDefault.prototype, "inputClass", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], EditCellDefault.prototype, "edited", void 0);
    EditCellDefault = __decorate([
        Component({
            template: ''
        })
    ], EditCellDefault);
    return EditCellDefault;
}());
export { EditCellDefault };
