// ng2-smart-table/esm5/lib/components/filter/custom-filter.component.js
import { __decorate, __extends, __metadata } from "tslib";
import { Component, ComponentFactoryResolver, Input, OnChanges, OnDestroy, SimpleChanges, ViewChild, ViewContainerRef } from '@angular/core';
import { FilterDefault } from './filter-default';
var CustomFilterComponent = /** @class */ (function (_super) {
    __extends(CustomFilterComponent, _super);
    function CustomFilterComponent(resolver) {
        var _this = _super.call(this) || this;
        _this.resolver = resolver;
        return _this;
    }
    CustomFilterComponent.prototype.ngOnChanges = function (changes) {
        var _this = this;
        if (this.column && !this.customComponent) {
            var componentFactory = this.resolver.resolveComponentFactory(this.column.filter.component);
            this.customComponent = this.dynamicTarget.createComponent(componentFactory);
            // set @Inputs and @Outputs of custom component
            this.customComponent.instance.query = this.query;
            this.customComponent.instance.column = this.column;
            this.customComponent.instance.source = this.source;
            this.customComponent.instance.inputClass = this.inputClass;
            this.customComponent.instance.filter.subscribe(function (event) { return _this.onFilter(event); });
        }
        if (this.customComponent) {
            this.customComponent.instance.ngOnChanges(changes);
        }
    };
    CustomFilterComponent.prototype.ngOnDestroy = function () {
        if (this.customComponent) {
            this.customComponent.destroy();
        }
    };
    CustomFilterComponent.ctorParameters = function () { return [
        { type: ComponentFactoryResolver }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], CustomFilterComponent.prototype, "query", void 0);
    __decorate([
        ViewChild('dynamicTarget', { read: ViewContainerRef, static: true }),
        __metadata("design:type", Object)
    ], CustomFilterComponent.prototype, "dynamicTarget", void 0);
    CustomFilterComponent = __decorate([
        Component({
            selector: 'custom-table-filter',
            template: "<ng-template #dynamicTarget></ng-template>"
        }),
        __metadata("design:paramtypes", [ComponentFactoryResolver])
    ], CustomFilterComponent);
    return CustomFilterComponent;
}(FilterDefault));
export { CustomFilterComponent };
