// ng2-smart-table/esm2015/lib/components/filter/filter-default.js
import { __decorate, __metadata } from "tslib";
import { Output, EventEmitter, Input, Component } from '@angular/core';
import { Column } from '../../lib/data-set/column';
import { DataSource } from '../../lib/data-source/data-source';
let FilterDefault = class FilterDefault {
    constructor() {
        this.inputClass = '';
        this.filter = new EventEmitter();
        this.query = '';
    }
    onFilter(query) {
        this.source.addFilter({
            field: this.column.id,
            search: query,
            filter: this.column.getFilterFunction(),
        });
    }
};
__decorate([
    Input(),
    __metadata("design:type", Column)
], FilterDefault.prototype, "column", void 0);
__decorate([
    Input(),
    __metadata("design:type", DataSource)
], FilterDefault.prototype, "source", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], FilterDefault.prototype, "inputClass", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], FilterDefault.prototype, "filter", void 0);
FilterDefault = __decorate([
    Component({
        template: ''
    })
], FilterDefault);
export { FilterDefault };
